### PR TITLE
fix uninitialized read when stringifying an addrLocal

### PR DESCRIPTION
**Pull request by kazcw**
Reachable from either place where SetIP is used when all of:
- our best-guess addrLocal for a peer is IPv4
- the peer tells us it's reaching us at an IPv6 address
- NET logging is enabled

In that case, SetIP turns an IPv4 address into an IPv6 address without
setting the scopeId, which is subsequently read in GetSockAddr during
CNetAddr::ToStringIP and passed to getnameinfo. Fix by ensuring every
constructor initializes the scopeId field with something.

### CODE REVIEW

**Comment by sipa** [comment]
Nice catch. Can you add a test that tests for the old behavior?

File: src/netaddress.cpp [hunk 1]
```diff
@@ -70,6 +70,7 @@ bool CNetAddr::SetSpecial(const std::string &strName)
 CNetAddr::CNetAddr(const struct in_addr& ipv4Addr)
 {
     SetRaw(NET_IPV4, (const uint8_t*)&ipv4Addr);
+    scopeId = 0;
```
Comment (promag) [inline src/netaddress.cpp]: I'd go a step further and move `scopeId = 0` to static initialization instead:
```
    uint32_t scopeId{0}; // for scoped/link-local ipv6 addresses
```

**Review by promag — COMMENTED** [review]
utACK 9151f78.

**Comment by practicalswift** [comment]
Excellent catch and thanks for another great contribution!

utACK 9151f783ca53cde718a3895324bd6389e5de6318 modulo @promag 

How did you find this issue?

I can't believe I didn't catch this issue as part of my routine static analysis runs :-\  I checked now and I had this specific case in multiple analysis reports, but it appears the signal drowned in the noise.

Would be nice to have the offending code path executed in a test to allow for catching it via dynamic analysis too.

File: src/netaddress.cpp (hunk 1 above)
Comment (kazcw) [reply src/netaddress.cpp]: Good point! Switched to in-class initialization.

**Comment by practicalswift** [comment]
utACK 04ae6807bc09d52c4738836243cb3dad8845b405

File: src/test/net_tests.cpp [hunk 2]
```diff
@@ -189,4 +189,42 @@ BOOST_AUTO_TEST_CASE(cnode_simple_test)
...
+    in6_addr ipv6AddrLocal;
+    memset(ipv6AddrLocal.s6_addr, 0, 16);
+    ipv6AddrLocal.s6_addr[0] = 0xcc;
+    CAddress addrLocal = CAddress(CService(ipv6AddrLocal, 7777), NODE_NETWORK);
+    pnode->SetAddrLocal(addrLocal);
+
+    // before patch, this causes undefined behavior detectable with clang's -fsanitize=memory
+    AdvertiseLocal(&*pnode);
+
+    // suppress no-checks-run warning; if this test fails, it's by triggering triggering a sanitizer
```
Comment (practicalswift) [inline src/test/net_tests.cpp]: Double "triggering"? :-)

**Comment by practicalswift** [comment]
utACK d0b673176b3214e16e4df50dcb65352fbd07969f modulo typo fix :-)

**Comment by promag** [comment]
utACK b7b36de.

**Comment by practicalswift** [comment]
utACK 8ebbef016928811756e46b9086067d1c826797a8

**Comment by MarcoFalke** [comment]
utACK b7b36decaf878a8c1dcfdb4a27196c730043474b

Makes the code adhere to our developer cpp style guide and also fixes the warning that valgrind yells at me when running `valgrind ./src/test/test_bitcoin -t net_tests/ipv4_peer_with_ipv6_addrMe_test`.

**Review by wailo — CHANGES_REQUESTED** [review]
I would suggest applying the same change to the ip array ```unsigned char ip[16] = {0};``` and remove the memset call from the constructor. This leaves an empty constructor which can then be marked default.

**Comment by laanwj** [comment]
utACK b7b36decaf878a8c1dcfdb4a27196c730043474b

@wailo's comment could be done later, but is not related to the fix for UB here

**Comment by practicalswift** [comment]
re-utACK 8ebbef016928811756e46b9086067d1c826797a8

**Comment by laanwj** [comment]
> re-utACK 8ebbef0

You're not acking the top commit (b7b36decaf878a8c1dcfdb4a27196c730043474b) is this intentional?

**Comment by fanquake** [comment]
Backported in #14835.

**Pull request merged**